File: monotonicStack/7.removeDuplicateLetters/js/removeDuplicateLetters.js
/**
 * @param {string} s
 * @return {string}
 */
var removeDuplicateLetters = function (s) {
    // 创建一个栈
    const stack = []
    // 创建一个数组，用来记录每个字母出现的次数
    const quantity = new Array(26).fill(0)
    // 创建一个数组，用来记录每个字母是否被使用
    const used = new Array(26).fill(false)
    // 遍历字符串，记录每个字母出现的次数
    for (let i of s) {
        quantity[i.charCodeAt(0) - 97]++
    }
    // 遍历字符串
    for (let i of s) {
        // 如果该字母已被使用，则跳过
        if (used[i.charCodeAt(0) - 97]) {
            quantity[i.charCodeAt(0) - 97]--
            continue
        }
        // 如果栈不为空，且栈顶元素大于当前元素，且栈顶元素在字符串中仍存在，则弹出栈顶元素
        while (stack.length > 0 && stack.at(-1) > i && quantity[stack.at(-1).charCodeAt(0) - 97] > 0) {
            used[stack.at(-1).charCodeAt(0) - 97] = false
            stack.pop()
        }
        // 将当前元素压入栈中
        stack.push(i)
        // 标记当前元素已被使用
        used[i.charCodeAt(0) - 97] = true
        // 将当前元素出现的次数减一
        quantity[i.charCodeAt(0) - 97]--
    }
    // 返回栈中的元素组成的字符串
    return stack.join("")
};

File: monotonicStack/7.removeDuplicateLetters/py/removeDuplicateLetters.py
# 定义一个函数，用于移除字符串中的重复字母
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        # 定义一个栈，用于存储最终结果
        stack = []
        # 定义一个列表，用于存储每个字母出现的次数
        quantity = [0 for _ in range(26)]
        # 定义一个列表，用于存储每个字母是否被使用
        used = [False for _ in range(26)]
        # 遍历字符串中的每一个字母
        for i in s:
            # 记录每个字母出现的次数
            quantity[ord(i) - ord('a')] += 1
        # 遍历字符串中的每一个字母
        for i in s:
            # 如果该字母已被使用，则跳过
            if used[ord(i) - ord('a')]:
                quantity[ord(i) - ord('a')] -= 1
                continue
            # 如果栈不为空，且栈顶字母大于当前字母，且栈顶字母出现的次数大于0，则弹出栈顶字母
            while stack and i < stack[-1] and quantity[ord(stack[-1]) - ord('a')]:
                used[ord(stack[-1]) - ord('a')] = False
                stack.pop()
            # 将当前字母压入栈中
            stack.append(i)
            # 标记当前字母已被使用
            used[ord(i) - ord('a')] = True
            # 记录当前字母出现的次数
            quantity[ord(i) - ord('a')] -= 1
        # 返回最终结果
        return ''.join(stack)